finality-aleph/src/testing/data_store.rs

```diff
@@ -0,0 +1,645 @@
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_and_import_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+
+    for i in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..(i as usize)].to_vec();
+        let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+        test_handler.send_data(test_data.clone());
+    }
+    assert!(
+        timeout(TIMEOUT_FAIL, test_handler.next_block_request())
+            .await
+            .is_err(),
+        "Data Store is sending block requests without reason"
+    );
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn message_with_genesis_block_does_not_get_through() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let _blocks = test_handler
+        .build_and_import_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+
+    let test_data: TestData = vec![aleph_data_from_headers(vec![test_handler.get_header_at(0)])];
+    test_handler.send_data(test_data.clone());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with genesis block proposal")
+        .await;
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn unimported_hopeless_forks_go_through() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10)
+        .await;
+    let forking_block = &blocks[MAX_DATA_BRANCH_LEN + 2];
+    let forks = test_handler
+        .build_branch_upon(&forking_block.hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    for i in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = forks[0..(i as usize)].to_vec();
+        let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+        test_handler.send_data(test_data.clone());
+    }
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.import_branch(blocks.clone(), false).await;
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 1].hash());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN * 2 + 1].hash());
```

Comment:
is `MAX_DATA_BRANCH_LEN + 2` sufficient?

Reply:
It would be not enough as this is the "base" of the fork, and the proposal starts one block above. But `MAX_DATA_BRANCH_LEN + 3` is the smallest number that is enough. I changed the test to reflect that.